doc/synchronization.md

```diff
@@ -0,0 +1,356 @@
+    meaning as in Java.)
+-   **atomicity (A)** - Operation is either done or not as a whole.
+-   **serialized (S)** - Operations are serialized in some order (they
+    cannot disappear). This is a new property not mentioned in other memory
+    models, since Java and C++ do not have dynamically defined fields. All
+    operations on one line in the table are serialized with each other.
+
+### Core behavior:
+
+| Operation | V | A | S | Notes |
+|:----------|:-:|:-:|:-:|:-----|
+| local variable read/write/definition | - | x | x | Local variables are determined during parsing, they are not usually dinamically added (with exception of `local_variable_set`). Therefore definition is quite rare. |
+| instance variable read/write/(un)definition | - | x | x ||
+| class variable read/write/(un)definition | x | x | x ||
+| global variable read/write/definition | x | x | x | un-define us not possible currently. |
+| constant variable read/write/(un)definition | x | x | x ||
+| `Thread` local variable read/write/definition | - | x | x | un-define us not possible currently. |
+| `Fiber` local variable read/write/definition | - | x | x | un-define us not possible currently. |
+| method creation/redefinition/removal | x | x | x ||
+| include/extend | x | x | x | If `AClass` is included `AModule`, `AClass` gets all `AModule`'s methods at once. |
+
+
+Notes:
+
+-   Variable read reads value from preexisting variable.
+-   Variable definition creates new variable (operation is serialized with
+    writes, implies an update cannot be lost).
+-   Module/Class definition is actually constant definition. It is defined
+    instantly, however its methods are then processed sequentially.
+-   `||=`, `+=`, etc. are actually two operations read and write which implies
+    that it's not an atomic operation. See volatile variables
+    with compare-and-set.
+-   Method invocation does not have any special properties that includes
+    object initialization.
+
+Implementation differences from the model:
+
+-   MRI: everything is volatile.
+-   JRuby: `Thread` and `Fiber` local variables are volatile. Instance
+    variables are volatile on x86 and people may un/intentionally depend
+    on the fact.
+-   Class variables require investigation.
+
+### Source loading:
+
+| Operation | V | A | S | Notes |
+|:----------|:-:|:-:|:-:|:-----|
+| requiring | x | x | x | File will not be required twice, classes and modules are still defined gradually. |
+| autoload | x | x | x | Only one autoload at a time. |
+
+Notes:
+
+-   Beware of requiring and autoloading in concurrent programs, it's possible to
+    see partially defined classes. Eager loading or blocking until class is
+    fully loaded has should be used to mitigate.
+
+### Core classes
+
+`Mutex`, `Monitor`, `Queue` have to work correctly on each implementation. Ruby
+implementations should not crash when e.g. Array is used in parallel environment
+but it may loose updates etc.
```

Comment:
I think this is absolutely the right decision - but I think it needs some narrative to justify. We do not syncohronise Array etc on your behalf. You need to use locks on them when you use them concurrently, but this means you do not.

Also, what does crash mean? JRuby will give you concurrent modification exceptions, which are normal Ruby exceptions. I think Rubinius might as well. Does that count as a crash?